internal/expr/parser.go

```diff
@@ -486,48 +512,76 @@ func (p *Parser) skipEnclosedParentheses() (bool, error) {
 //  - bool == false, err == nil
 //		The construct was not the one we are looking for
 
-// parseIdentifierAsterisk parses a name made up of only nameBytes or of a
-// single asterisk. On success it returns the parsed string and true. Otherwise,
-// it returns the empty string and false.
-func (p *Parser) parseIdentifierAsterisk() (string, bool) {
-	if p.skipByte('*') {
-		return "*", true
+// parseColumnName parses an identifier or an asterisk.
+func (p *Parser) parseColumnName() (string, bool, error) {
+	if p.skipChar('*') {
+		return "*", true, nil
 	}
-	return p.parseIdentifier()
+
+	mark := p.pos
+
+	// parse quoted column names.
+	if ok, err := p.skipStringLiteral(); err != nil {
+		return "", false, err
+	} else if ok {
+		return p.input[mark:p.pos], true, nil
+	}
+
+	// parse regular column names, including numeric literals.
+	for p.pos < len(p.input) && isNameChar(p.char) {
+		p.advanceChar()
+	}
+
+	if p.pos > mark {
+		return p.input[mark:p.pos], true, nil
+	}
+	return "", false, nil
 }
 
-// parseIdentifier parses a name made up of only nameBytes. On success it
-// returns the parsed string and true. Otherwise, it returns the empty string
-// and false.
-func (p *Parser) parseIdentifier() (string, bool) {
+// parseTypeName parses a name starting with a letter or underscore and
+// followed by letters, digits and underscores. This matches the allowed
+// characters in Go type names.
+func (p *Parser) parseTypeName() (string, bool) {
 	mark := p.pos
-	if p.skipName() {
+
+	if isInitialNameChar(p.char) {
+		p.advanceChar()
+		for p.pos < len(p.input) && isNameChar(p.char) {
+			p.advanceChar()
+		}
+	}
+
+	if p.pos > mark {
 		return p.input[mark:p.pos], true
 	}
 	return "", false
 }
 
-// parseColumnAccessor parses either a column made up of name bytes optionally
+// parseColumnAccessor parses either a column made up of name chars optionally
 // dot-prefixed by its table name or a SQL function call used in place of a
 // column.
 // parseColumnAccessor returns an error so that it can be used with parseList.
 func (p *Parser) parseColumnAccessor() (columnAccessor, bool, error) {
 	cp := p.save()
 
 	// asterisk cannot be followed by anything.
-	if p.skipByte('*') {
+	if p.skipChar('*') {
 		return basicColumn{column: "*"}, true, nil
 	}
 
-	id, ok := p.parseIdentifier()
+	// Parse a column or a table name.
+	id, ok, err := p.parseColumnName()
```

Comment:
Changed the name back to `parseIdentifer` and reinstate `parseIdentifierAsterisk`. Identifier matches the SQL column and table names as well as what we accept for db tags, since this is the same thing. A generic name like Identifier makes more sense.